### ISSUE TITLE
Godot 4

### ISSUE THREAD

**Issue by TheBlckbird** [hiulit/Unofficial-Godot-Engine-Raspberry-Pi#29]
Will you compile Godot 4 for the Raspberry Pi?

**Comment by hiulit** [comment]
I will when (if ever) it's compatible again with GLES2. Right now it won't work.

**Comment by TheBlckbird** [comment]
I always used GLES3 and it worked perfectlyViele Grüße,Louis WeigelAm 03.03.2023 um 11:27 schrieb hiulit ***@***.***>:﻿
I will when (if ever) it's compatible again with GLES2. Right now it won't work.

**Comment by OverloadedOrama** [comment]
[Raspberry Pi 4 has had support for Vulkan 1.2 since August 2022](https://www.raspberrypi.com/news/vulkan-update-version-1-2-conformance-for-raspberry-pi-4/), so Godot 4's Vulkan renderers *should* work with the Raspberry Pi 4, since they require Vulkan 1.0 minimum. Although it may be worth it to wait for Godot 4.1 for performance improvements, and I imagine it will not work at all with older Raspberry Pis.

**Comment by hiulit** [comment]
@TheBlckbird How are running GLES3 on the Pi? It never worked for me

**Comment by TheBlckbird** [comment]
@hiulit idk it just works but the the text saying GLES3 where I can select whether I wanna use it or GLES2 is red until I change it to GLES2

**Comment by TheBlckbird** [comment]
And I also tried compiling Godot 4 on my own with this command:
`scons`
but when launching the compiled binary, I get this error message:
```

Error: Couldn't load project data at path ".". Is the .pck file missing? If you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension).
```

**Comment by frankhale** [comment]
@TheBlckbird looks like you compiled Godot without the editor. I believe you have to pass `target=editor` to scons when compiling.

**Comment by TheBlckbird** [comment]
Oh wait I just realized that I compiled the `target=template_release`

**Comment by hiulit** [comment]
I've compiled Godot 4.0-stable and it crashes :(
I'll have to investigate.

**Comment by TheBlckbird** [comment]
Same for me

**Comment by hiulit** [comment]
I've "successfully" launched Godot 4, but using software rendering `"LIBGL_ALWAYS_SOFTWARE=1" ./godot..blahblah`

I've installed the latest mesa drivers and it still crashes using both Vulkan and OpenGL.

**Comment by jmcerrejon** [comment]
> I've "successfully" launched Godot 4, but using software rendering `"LIBGL_ALWAYS_SOFTWARE=1" ./godot..blahblah`
> 
> I've installed the latest mesa drivers and it still crashes using both Vulkan and OpenGL.

Are you using the right parameters? A Mesa developer contributed to my script that installs Vulkan Driver on the Pi, and maybe those improvements can make our dream come true. Check it out [here](https://github.com/jmcerrejon/PiKISS/commit/b94c768c3e7b8f0240da4cea7734001fbc98e5b2).

**Comment by Pedrocasf** [comment]
So i have successfully launched Godot 4.0.2 on the Raspberry Pi 4, by adding two small changes to `drivers/vulkan/rendering_device_vulkan.cpp` .First on the first line of the function with the signature 

`VmaPool RenderingDeviceVulkan::_find_or_create_small_allocs_pool(uint32_t p_mem_type_index) `
add the following line  
`p_mem_type_index = p_mem_type_index == std::numeric_limits<uint32_t>::max() ? 0 : p_mem_type_index;`
second chnage the line 
`static const uint32_t SMALL_ALLOCATION_MAX_SIZE = 4096;` 
to 
`static const uint32_t SMALL_ALLOCATION_MAX_SIZE = 0xFFFFFFFE;`
or just a really large number that is less than 
` std::numeric_limits<uint32_t>::max()`
the when launching pass the args `--display-driver x11 --rendering-driver vulkan`
PS: i had also compiled the latest mesa driver source

**Comment by hiulit** [comment]
> > I've "successfully" launched Godot 4, but using software rendering `"LIBGL_ALWAYS_SOFTWARE=1" ./godot..blahblah`
> > I've installed the latest mesa drivers and it still crashes using both Vulkan and OpenGL.
> 
> Are you using the right parameters? A Mesa developer contributed to my script that installs Vulkan Driver on the Pi, and maybe those improvements can make our dream come true. Check it out [here](https://github.com/jmcerrejon/PiKISS/commit/b94c768c3e7b8f0240da4cea7734001fbc98e5b2).

Thanks! I'll give it another try when I have the time. I actually used you scripts to install the Vulkan drivers on the Pi ;)

**Comment by hiulit** [comment]
> So i have successfully launched Godot 4.0.2 on the Raspberry Pi 4, by adding two small changes to `drivers/vulkan/rendering_device_vulkan.cpp` .First on the first line of the function with the signature
> 
> `VmaPool RenderingDeviceVulkan::_find_or_create_small_allocs_pool(uint32_t p_mem_type_index) ` add the following line `p_mem_type_index = p_mem_type_index == std::numeric_limits<uint32_t>::max() ? 0 : p_mem_type_index;` second chnage the line `static const uint32_t SMALL_ALLOCATION_MAX_SIZE = 4096;` to `static const uint32_t SMALL_ALLOCATION_MAX_SIZE = 0xFFFFFFFE;` or just a really large number that is less than ` std::numeric_limits<uint32_t>::max()` the when launching pass the args `--display-driver x11 --rendering-driver vulkan` PS: i had also compiled the latest mesa driver source

That's good news! Maybe you could open an issue on the Godot Github and see if they are interested in those changes.

**Comment by Pedrocasf** [comment]
Second thing, using the latest mesa drivers thatt support OpenGL 3.1 it is possible to set MESA_GL_VERSION_OVERRIDE=3.3 and MESA_GLSL_VERSION_OVERRIDE=330 and it "should" work because Godot will probably only use features from OpenGL-ES 3.0 .

**Comment by dolfelt** [comment]
@Pedrocasf Could you describe how you got it to build? I'm interested in getting an export template for a Raspberry Pi4 with Godot 4.1. Right now there is very little information or forks or anything on how to build it for arm.

Any help would be appreciated!

**Comment by Pedrocasf** [comment]
@dolfelt, so, I only did the aforementioned changes and folowed the godot 4 docs [Compiling for Linux-BSD](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_linuxbsd.html).
Also a few issues you might find. First when building on arm64 dont use gcc with full LTO, at least on the latest commit of both 4.x and 3.x, as it uses more than 8 GiB of RAM to do the LTO. Second, Compile the latest Mesa dev commit, it seems to be more stable than the current official release. [Q-engineering](https://qengineering.eu/install-vulkan-on-raspberry-pi.html) has this guide on how to do that, altough it is outdated.

**Comment by damvcoool** [comment]
At this point I would be happy with just Export Template for Linux aarch64.

**Comment by hiulit** [comment]
Hi, everybody!

First of all, great news, because we have official ARM binaries from the Godot team https://downloads.tuxfamily.org/godotengine/testing/4.2-beta-2023-11-01-linux-arm/

They also released an arm64 toolchain to easliy cross-compile Godot. I'll update my cross-compilation project https://github.com/hiulit/cross-compile-godot-raspberry-pi to use it

So, here's a [recap of my latest tests on Godot 4](https://x.com/hiulit/status/1722953726963429380?s=20). I used the latest Pi 5.
Apparently, the Raspberry Pi OS has issues with the Vulkan drivers, event with the latest ones, so I used Ubuntu 23.10.
Basically, Godot 4 works perfectly using `--rendering-driver opengl3_es` and `Compatibilty` graphics renderer.
It also seems to work using Vulkan, at leat it opens, on Ubuntu 23.10, but it keeps crashing afterwards.
